spark-jobs/src/main/scala/za/co/absa/enceladus/standardization/StandardizationJob.scala

```diff
@@ -424,53 +425,58 @@ object StandardizationJob {
     */
-  private def addRawRecordCountToMetadata(df: DataFrame): Unit = {
+  private def addRawAndSourceRecordCountsToMetadata: Unit = {
     val checkpointRawRecordCount = getRawRecordCountFromCheckpoints
+    val checkpointSourceRecordCount = getSourceRecordCountFromCheckpoints
 
-    val rawRecordCount = checkpointRawRecordCount match {
-      case Some(num) => num
-      case None      => df.count
+    (checkpointRawRecordCount, checkpointSourceRecordCount) match {
+      case (Right(x), Right(y)) => {
+        Atum.setAdditionalInfo(s"raw_record_count" -> x.toString)
+        Atum.setAdditionalInfo(s"source_record_count" -> y.toString)
+      }
+      case (Right(_), Left(er)) => throw new ValidationException("Field errors", Seq(er))
+      case (Left(er), Right(_)) => throw new ValidationException("Field errors", Seq(er))
+      case (Left(er1), Left(er2)) => throw new ValidationException("Field errors", Seq(er1,er2))
     }
-    Atum.setAdditionalInfo(s"raw_record_count" -> rawRecordCount.toString)
   }
 
   /**
     * Gets the number of records in raw data by traversing Atum's checkpoints.
     *
     * @return The number of records in a checkpoint corresponding to raw data (if available)
     */
-  private def getRawRecordCountFromCheckpoints: Option[Long] = {
+  private def getRawRecordCountFromCheckpoints: Either[String, Long] = {
+    getFieldRawRecordFromCheckpoints("row")
+  }
+
+  /**
+   * Gets the number of records in source data by traversing Atum's checkpoints.
+   *
+   * @return The number of records in a checkpoint corresponding to source data (if available)
+   */
+  private def getSourceRecordCountFromCheckpoints: Either[String, Long] = {
+    getFieldRawRecordFromCheckpoints("source")
+  }
+
+  private def getFieldRawRecordFromCheckpoints(field: String) = {
     import za.co.absa.atum.core.Constants._
-    val controlMeasure = Atum.getControlMeasure
-
-    val rawCheckpoint = controlMeasure
-      .checkpoints
-      .find(c => c.name.equalsIgnoreCase("raw") || c.workflowName.equalsIgnoreCase("raw"))
-
-    val measurement = rawCheckpoint.flatMap(chk => {
-      chk.controls.find(m => m.controlType.equalsIgnoreCase(controlTypeRecordCount))
-    })
-
-    measurement.flatMap(m =>
-      try {
-        val rawCount = m.controlValue.toString.toLong
-        // Basic sanity check
-        if (rawCount >= 0) {
-          Some(rawCount)
-        } else {
-          None
-        }
-      }
-      catch {
-        case NonFatal(_) => None
-      }
-    )
+    val optionResult = for {
+      checkpoint <- Atum.getControlMeasure
+        .checkpoints
+        .find(c => c.name.equalsIgnoreCase(field) || c.workflowName.equalsIgnoreCase(field))
+
+      measurement <- checkpoint.controls.find(m => m.controlType.equalsIgnoreCase(controlTypeRecordCount))
+      result <- Try({
+        val rowCount = measurement.controlValue.toString.toLong
+        if (rowCount >= 0) rowCount else throw new Exception("negative rowcount")
```

Comment:
Why not return `Left` here as well? `Option` might become unnecessary.

Reply:
I think I did that because I could not use Either in the for comprehension, but I may be wrong

Reply:
Yep, you a correct. I've missed that, sorry.